runtime/src/registrar.rs

```diff
@@ -0,0 +1,1346 @@
+	type Currency: ReservableCurrency<Self::AccountId>;
+
+	/// The deposit to be paid to run a parathread.
+	type ParathreadDeposit: Get<BalanceOf<Self>>;
+
+	/// Handler for when two ParaIds are swapped.
+	type OnSwap: OnSwap;
+}
+
+/// The number of items in the parathread queue, aka the number of blocks in advance to schedule
+/// parachain execution.
+const QUEUE_SIZE: usize = 2;
+
+/// The number of rotations that you will have as grace if you miss a block.
+const MAX_RETRIES: u32 = 3;
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Registrar {
+		// Ordered vector of all parachain IDs.
+		Parachains: Vec<ParaId>;
+
+		/// The number of threads to schedule per block.
+		ThreadCount: u32;
+
+		/// An array of the queue of set of threads scheduled for the coming blocks; ordered by
+		/// para ID. There can be no duplicates of para ID in each list item.
+		SelectedThreads: Vec<Vec<(ParaId, CollatorId)>>;
+
+		/// Parathreads/chains scheduled for execution this block. If the collator ID is set, then
+		/// a particular collator has already been chosen for the next block, and no other collator
+		/// may provide the block. In this case we allow the possibility of the combination being
+		/// retried in a later block, expressed by `Retriable`.
+		///
+		/// Ordered as Parachains ++ Selected_Parathreads.
+		Active: Vec<(ParaId, Option<(CollatorId, Retriable)>)>;
+
+		/// The next unused ParaId value. Start this high in order to keep low numbers for
+		/// system-level chains.
+		NextFreeId: ParaId = LOWEST_USER_ID;
+
+		/// Pending swap operations.
+		PendingSwap: map ParaId => Option<ParaId>;
+
+		/// Map of all registered parathreads/chains.
+		Paras get(paras): map ParaId => Option<ParaInfo>;
+
+		/// The current queue for parathreads that should be retried.
+		RetryQueue get(retry_queue): [Vec<(ParaId, CollatorId)>; MAX_RETRIES as usize];
+
+		/// Users who have paid a parathread's deposit
+		Debtors: map ParaId => T::AccountId;
+	}
+	add_extra_genesis {
+		config(parachains): Vec<(ParaId, Vec<u8>, Vec<u8>)>;
+		config(_phdata): PhantomData<T>;
+		build(build::<T>);
+	}
+}
+
+#[cfg(feature = "std")]
+fn build<T: Trait>(config: &GenesisConfig<T>) {
+	let mut p = config.parachains.clone();
+	p.sort_unstable_by_key(|&(ref id, _, _)| *id);
+	p.dedup_by_key(|&mut (ref id, _, _)| *id);
+
+	let only_ids: Vec<ParaId> = p.iter().map(|&(ref id, _, _)| id).cloned().collect();
+
+	Parachains::put(&only_ids);
+
+	for (id, code, genesis) in p {
+		Paras::insert(id, &primitives::parachain::PARACHAIN_INFO);
+		// no ingress -- a chain cannot be routed to until it is live.
+		<parachains::Code>::insert(&id, &code);
+		<parachains::Heads>::insert(&id, &genesis);
+		<parachains::Watermarks<T>>::insert(&id, T::BlockNumber::zero());
+		// Save initial parachains in registrar
+		Paras::insert(id, ParaInfo { scheduling: Scheduling::Always })
+	}
+}
+
+pub fn swap_ordered_existence<T: PartialOrd + Ord + Copy>(ids: &mut Vec<T>, one: T, other: T) {
```

Comment:
missing docs

Reply:
also should take `&[T]` - it doesn't seem that the slice being allocated on the heap or being dynamically sized is a pre-requisite of the interface